Kernel/Bus/USB/xHCI/xHCIController.cpp

```diff
@@ -0,0 +1,1439 @@
+    auto const slot = device.controller_identifier();
+
+    TransferRequestBlock transfer_request_blocks[3] {};
+    auto trb_index = 0u;
+
+    auto* setup_trb = &transfer_request_blocks[trb_index++];
+    memcpy(setup_trb, &transfer.request(), sizeof(USBRequestData));
+    setup_trb->setup_stage.transfer_request_block_transfer_length = 8; // Always 8.
+    setup_trb->setup_stage.interrupter_target = 0;
+    setup_trb->setup_stage.interrupt_on_completion = 0;
+    setup_trb->setup_stage.immediate_data = 1; // This bit shall be set to ‘1’ in a Setup Stage TRB
+    setup_trb->setup_stage.transfer_request_block_type = TransferRequestBlock::TRBType::Setup_Stage;
+    if (transfer.transfer_data_size() > 0)
+        setup_trb->setup_stage.transfer_type = direction_in ? TransferRequestBlock::TransferType::IN_Data_Stage : TransferRequestBlock::TransferType::OUT_Data_Stage;
+    else
+        setup_trb->setup_stage.transfer_type = TransferRequestBlock::TransferType::No_Data_Stage;
+
+    if (transfer.transfer_data_size() > 0) {
+        auto* data_trb = &transfer_request_blocks[trb_index++];
+        auto data_buffer_paddr = transfer.buffer_physical().get() + sizeof(USBRequestData); // FIXME: This is an ugly hack in the USB subsystem that works around a UHCI-specific issue, get rid of this
+        data_trb->data_stage.data_buffer_low = data_buffer_paddr;
+        data_trb->data_stage.data_buffer_high = data_buffer_paddr >> 32;
+        data_trb->data_stage.transfer_request_block_transfer_length = transfer.transfer_data_size();
+        data_trb->data_stage.transfer_descriptor_size = 0;
+        data_trb->data_stage.interrupter_target = 0;
+        data_trb->data_stage.chain_bit = 0;
+        data_trb->data_stage.interrupt_on_completion = 0;
+        data_trb->data_stage.immediate_data = 0;
+        data_trb->data_stage.transfer_request_block_type = TransferRequestBlock::TRBType::Data_Stage;
+        data_trb->data_stage.direction = direction_in;
+    }
+
+    auto* status_trb = &transfer_request_blocks[trb_index++];
+    status_trb->status_stage.interrupter_target = 0;
+    status_trb->status_stage.chain_bit = 0;
+    status_trb->status_stage.interrupt_on_completion = 1;
+    status_trb->status_stage.transfer_request_block_type = TransferRequestBlock::TRBType::Status_Stage;
+    status_trb->status_stage.direction = !direction_in || transfer.transfer_data_size() == 0;
+
+    SyncPendingTransfer pending_transfer;
+    TRY(enqueue_transfer(slot, 0, Pipe::Direction::Bidirectional, { transfer_request_blocks, trb_index }, pending_transfer));
+    pending_transfer.wait_queue.wait_forever();
+    VERIFY(!pending_transfer.endpoint_list_node.is_in_list());
+
+    if (pending_transfer.completion_code != TransferRequestBlock::CompletionCode::Success)
+        return EINVAL;
+
+    return transfer.transfer_data_size() - pending_transfer.remainder;
+}
+
+ErrorOr<size_t> xHCIController::submit_bulk_transfer(Transfer&)
+{
+    TODO();
```

Comment:
I don't like just returning those errors, because then we fail silently due to known missing functionality. I'll make it return an error, but also dbgln an appropriate TODO